content/blog/2025-12-16-tiling-fusion-search.md

```diff
@@ -0,0 +1,135 @@
+For validity, ensure that for each dimension in the original einsum, the product of its factors in the tiled loop nest is at least its total size. To ensure we don't overflow the cache, the sum of buffer sizes in the cache should be at most a given capacity.
+
+So how can we find a tiled loop nest of optimal performance? As a first attempt at expressing the optimal tiling problem using CP-SAT, let’s try splitting each dimension into a fixed number of tiled loops (say 2) and then let the solver choose:
+1. The order of the factored loops
+2. The factor size in each loop
+3. The blocking level of each tensor
+This works fine; the positions of the loops and blocking levels can be represented using boolean variables. For a 6-loop schedule of matmul, for instance, we will have 36 boolean variables of the form "loop __i__ is assigned to position __j__" and 18 boolean variables of the form "position __j__ is inside the blocking level for tensor A".
+
+Note that this has some redundancy, though. We’re asking the solver to choose the exact position of every loop, but we only care whether each loop is inside or outside each blocking level. To get rid of this, we can try asking the solver to instead choose only a partial order over the loops, from which we can later extract any loop schedule consistent with that partial order. This formulation is available [here](https://github.com/natetyoung/tensor-program-dse/blob/main/po_einsum_cpsat97.py).
+
+This is still a little redundant, though. We’re not making the solver choose a total order, but it's still _allowed_ to choose one, so it might be spending extra time proving to itself that the exact order doesn't matter. Another issue is that we have a few poorly-justified assumptions in our model. Why 2 factors of each dim specifically? Can we assume that we only have 1 factor for a reduction dimension (this makes the solver considerably faster, but is only justified by empirics).
+
+We can take our approach further in order to solve both of these problems at once. If all that matters is whether each factor is inside or outside each tiling level, let's _only choose the order of the tiling levels_ and then choose how much of each dim goes into each "bucket" created between them. In the tiled matmul example above, for instance, we have 4 buckets. The outermost contains factors `M/128` and `N/128` of `M` and `N` respectively; the second bucket has all of dim `K`; the third has factor `128` of `M`, and the innermost bucket has factor `128` of `N`. This formulation is available [here](https://github.com/natetyoung/tensor-program-dse/blob/main/cb_einsum_cpsat97.py).
```

Comment:
I still think this could use some further explanation to get clear about how the encoding works. I understand the reference to the example, but I don't yet see a direct description of which constraints need to be constructed to reflect these "buckets." What does it mean to "only choose the order of the tiling levels"? I think some step-by-step development of some example constraints could help a lot…